core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala

```diff
@@ -339,14 +339,17 @@ class DeltaSqlAstBuilder extends DeltaSqlBaseBaseVisitor[AnyRef] {
    * }}}
    */
   override def visitOptimizeTable(ctx: OptimizeTableContext): AnyRef = withOrigin(ctx) {
-    if (ctx.path == null && ctx.table == null) {
-      throw new ParseException("OPTIMIZE command requires a file path or table name.", ctx)
-    }
+    val path = Option(ctx.path).map(string)
+    val table = Option(ctx.table).map(visitTableIdentifier)
     val interleaveBy = Option(ctx.zorderSpec).map(visitZorderSpec).getOrElse(Seq.empty)
-    OptimizeTableCommand(
-      Option(ctx.path).map(string),
-      Option(ctx.table).map(visitTableIdentifier),
-      Option(ctx.partitionPredicate).map(extractRawText(_)).toSeq, Map.empty)(interleaveBy)
+    val predicates = Option(ctx.partitionPredicate).map(extractRawText(_)).toSeq
+    val plan = (path, table) match {
+      case(Some(p), None) => UnresolvedPathBasedDeltaTable(p, "OPTIMIZE")
+      case(None, Some(t)) => UnresolvedTable(t.database.toSeq :+ t.identifier, "OPTIMIZE", None)
+      case _ => throw new ParseException(
+        "OPTIMIZE command requires a file path or table name.", ctx)
+    }
+    OptimizeTableCommand(plan, predicates, Map.empty)(interleaveBy)
```

Comment:
Instead of changing the SQL parser and churning all the unit tests, can we declare an alternate constructor for `OptimizeTableCommand` that takes the path/table pair like today? Then as we fix other similar Delta commands, we should be able to factor out a helper method to do it once.

We'd just need to add some new unit tests to specifically verify that the command handles `UnresolvedPathBasedDeltaTable`, `UnresolvedTable`, and `ResolvedTable` correctly (with the latter obtained from a dataframe).